lucene/analysis/icu/src/tools/java/org/apache/lucene/analysis/icu/GenerateUTR30DataFiles.java

```diff
@@ -0,0 +1,274 @@
+      Writer writer = new OutputStreamWriter(out, StandardCharsets.UTF_8);
+      try {
+        writer.write(builder.toString());
+      } finally {
+        writer.close();
+      }
+    }
+  }
+
+  private static void getNFKCDataFilesFromIcuProject() throws IOException {
+    URL icuTagsURL = new URL(ICU_GIT_TAG_URL + "/");
+    URL icuReleaseTagURL = new URL(icuTagsURL, ICU_RELEASE_TAG + "/");
+    URL norm2url = new URL(icuReleaseTagURL, ICU_DATA_NORM2_PATH + "/");
+
+    System.err.print("Downloading " + NFKC_TXT + " ... ");
+    download(new URL(norm2url, NFKC_TXT), NFKC_TXT);
+    System.err.println("done.");
+    System.err.print("Downloading " + NFKC_CF_TXT + " ... ");
+    download(new URL(norm2url, NFKC_CF_TXT), NFKC_CF_TXT);
+    System.err.println("done.");
+
+    System.err.print("Downloading " + NFKC_CF_TXT + " and making diacritic rules one-way ... ");
+    URLConnection connection = openConnection(new URL(norm2url, NFC_TXT));
+    BufferedReader reader = new BufferedReader
+        (new InputStreamReader(connection.getInputStream(), StandardCharsets.UTF_8));
+    Writer writer = new OutputStreamWriter(new FileOutputStream(NFC_TXT), StandardCharsets.UTF_8);
+    try {
+      String line;
+
+      while (null != (line = reader.readLine())) {
+        Matcher matcher = ROUND_TRIP_MAPPING_LINE_PATTERN.matcher(line);
+        if (matcher.matches()) {
+          final String leftHandSide = matcher.group(1);
+          final String rightHandSide = matcher.group(2).trim();
+          List<String> diacritics = new ArrayList<>();
+          for (String outputCodePoint : rightHandSide.split("\\s+")) {
+            int ch = Integer.parseInt(outputCodePoint, 16);
+            if (UCharacter.hasBinaryProperty(ch, UProperty.DIACRITIC)
+                // gennorm2 fails if U+0653-U+0656 are included in round-trip mappings
+                || (ch >= 0x653 && ch <= 0x656)) {
+              diacritics.add(outputCodePoint);
+            }
+          }
+          if ( ! diacritics.isEmpty()) {
+            StringBuilder replacementLine = new StringBuilder();
+            replacementLine.append(leftHandSide).append(">").append(rightHandSide);
+            replacementLine.append("  # one-way: diacritic");
+            if (diacritics.size() > 1) {
+              replacementLine.append("s");
+            }
+            for (String diacritic : diacritics) {
+              replacementLine.append(" ").append(diacritic);
+            }
+            line = replacementLine.toString();
+          }
+        }
+        writer.write(line);
+        writer.write("\n");
+      }
+    } finally {
+      reader.close();
+      writer.close();
+    }
+    System.err.println("done.");
+  }
+
+  private static void download(URL url, String outputFile)
+      throws IOException {
+    final URLConnection connection = openConnection(url);
+    final InputStream inputStream = connection.getInputStream();
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code contains a resource that might not be closed properly. Resource leaks can cause your system to slow down or crash.

**Fix**
Consider closing the following resource: *inputStream*. The resource is referenced in statements at the following line: 230. The resource closure statement is at line: 234. There are other execution paths that do not contain closure statements, e.g., when exception is thrown by FileOutputStream constructor. Either a) close *inputStream* in a try-finally block or b) close the resource by declaring *inputStream* in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).